p2p/peer.go

```diff
@@ -214,97 +214,280 @@ type PeerUpdate struct {
+// for dialing again when appropriate.
+func (m *peerManager) DialFailed(peerID NodeID, address PeerAddress) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	delete(m.dialing, peerID)
+	// FIXME: We need to track address quality statistics and exponential backoff.
+	return nil
+}
+
+// Dialed marks a peer as successfully dialed. Any further incoming connections
+// will be rejected, and once disconnected the peer may be dialed again.
+func (m *peerManager) Dialed(peerID NodeID, address PeerAddress) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	peer, err := m.store.Get(peerID)
+	if err != nil {
 		return err
+	} else if peer == nil {
+		return fmt.Errorf("unknown peer %q", peerID)
 	}
-	peerID := address.NodeID()
 
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
+	if m.connected[peerID] {
+		return fmt.Errorf("peer %v is already connected", peerID)
+	}
+	delete(m.dialing, peerID)
+	m.connected[peerID] = true
+	return nil
+}
 
-	peer, ok := s.peers[peerID]
-	if !ok {
-		peer = newStorePeer(peerID)
-		s.peers[peerID] = peer
-	} else if s.claimed[peerID] {
-		// FIXME: We need to handle modifications of claimed peers somehow.
-		return fmt.Errorf("peer %q is claimed", peerID)
+// Accepted marks an incoming peer connection successfully accepted. If the peer
+// is already connected this will return an error.
+//
+// NOTE: We can't take an address here, since e.g. TCP uses a different port
+// number for outbound traffic than inbound traffic, so the peer's endpoint
+// wouldn't necessarily be an appropriate address to dial.
+func (m *peerManager) Accepted(peerID NodeID) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	peer, err := m.store.Get(peerID)
+	if err != nil {
+		return err
+	} else if peer == nil {
+		peer = newPeerInfo(peerID)
+		if err = m.store.Set(peer); err != nil {
+			return err
+		}
+	}
+	if m.connected[peerID] {
+		return fmt.Errorf("peer %q is already connected", peerID)
 	}
-	peer.AddAddress(address)
+	m.connected[peerID] = true
 	return nil
 }
 
-// Claim claims a peer. The caller has exclusive ownership of the peer, and must
-// return it by calling Return(). Returns nil if the peer could not be claimed.
-// If the peer is not known to the store, it is registered and claimed.
-func (s *peerStore) Claim(id NodeID) *peerInfo {
-	s.mtx.Lock()
-	defer s.mtx.Unlock()
-	if s.claimed[id] {
-		return nil
+// Ready marks a peer as ready, broadcasting status updates to subscribers. The
+// peer must already be marked as connected. This is separate from Dialed() and
+// Accepted() to allow the router to set up its internal queues before reactors
+// start sending messages (holding the Router.peerMtx mutex while calling
+// Accepted or Dialed will halt all message routing while peers are set up, which
+// is too expensive and also causes difficulties in tests where we may want to
+// consume peer updates and send messages sequentially).
+//
+// FIXME: This possibly indicates an architectural problem. Should the peerManager
```

Comment:
could you expand on this?

Reply:
`Ready()` exists only because the flow has to be like this:

1. Call `peerManager.Connected()`
2. Set up router queue
3. Call  `peerManager.Ready()` to trigger peer update

Otherwise, if we send the peer update during `Connected()` (which would be more logical), the reactor can send a message to the peer before the router has set up its queue, thus dropping that message. This detailed interplay between the manager and router suggests that the abstraction may be wrong, since they really shouldn't care too much about each other's internals. However, I don't think there's any way around it, and this approach is probably simpler to understand than e.g. introducing an `onConnected()` callback that the router could use for queue registration.